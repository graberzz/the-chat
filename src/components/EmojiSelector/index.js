import React from 'react';
import styled from 'styled-components';

const EMOJI_LIST = [...'😀😁😂😃😄😅😆😉😊😋😎😍😘😗😙😚🙂🤗🤔😐😑😶🙄😏😣😥😮🤐😯😪😫😴😌😛😜😒😓😔😕🙃🤑😲🙁😖😞😟😤😢😭😦😧😨😩😬😰😱😳😵😡😠😷🤒🤕😇🤓😈'];

const EmojiWrap = styled.div`
  position: relative;
`;

const EmojiToggle = styled.button`
  border: none;
  background: none;
  cursor: pointer;
  outline: none;
  font-size: 1.2em;
`;
EmojiToggle.displayName = 'EmojiToggle';

const EmojiList = styled.ul`
  position: absolute;
  right: -4px;
  bottom: 31px;
  display: flex;
  flex-wrap: wrap;
  width: 265px;
  background: #fff;
  font-size: 2em;
  user-select: none;
`;
EmojiList.displayName = 'EmojiList';

const EmojiItem = styled.li`
  cursor: pointer;
`;

class EmojiSelector extends React.Component {
  state = {
    open: false,
  }

  onToggleClick = () => this.setState({ open: !this.state.open })

  render() {
    const { open } = this.state;
    const { onEmojiClick } = this.props;
    return (
      <EmojiWrap>
        <EmojiToggle onClick={this.onToggleClick}>
          <span role='img' aria-label='toggle emoji'>😃</span>
        </EmojiToggle>
        {open && <EmojiList>
          {EMOJI_LIST.map(emoji =>
            <EmojiItem key={emoji} onClick={e => onEmojiClick(e.target.textContent)}>
              {emoji}
            </EmojiItem>)}
        </EmojiList>}
      </EmojiWrap>
    );
  }
}

export default EmojiSelector