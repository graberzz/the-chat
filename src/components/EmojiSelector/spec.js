import React from 'react';
import { shallow, mount } from 'enzyme';
import EmojiSelector from './';

describe('<EmojiSelector />', () => {
  let ES;

  beforeEach(() => {
    ES = shallow(<EmojiSelector />);
  });

  afterEach(() => {
    ES = undefined;
  });
  
  it('matches snapshot', () => {
    expect(ES).toMatchSnapshot();
  });

  it('does not render EmojiList initially', () => {    
    expect(ES.find('EmojiList')).toHaveLength(0);
    expect(ES.state().open).toBe(false);
  });


  it('toggle click renders EmojiList', () => {    
    ES.find('EmojiToggle').simulate('click');

    expect(ES.find('EmojiList')).toHaveLength(1);
    expect(ES.state().open).toBe(true);
  });

  it('onEmojiClick returns valid emoji', () => {
    const fn = jest.fn();
    ES = mount(<EmojiSelector onEmojiClick={fn}/>);
    ES.find('EmojiToggle').simulate('click');
    
    const firstEmoji = ES.find('li').at(0); 
    
    firstEmoji.simulate('click');

    expect(fn.mock.calls[0][0]).toBe(firstEmoji.text());
  });
});