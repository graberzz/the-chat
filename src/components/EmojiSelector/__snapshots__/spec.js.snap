// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmojiSelector /> matches snapshot 1`] = `
<styled.div>
  <EmojiToggle
    onClick={[Function]}
  >
    <span
      aria-label="toggle emoji"
      role="img"
    >
      ðŸ˜ƒ
    </span>
  </EmojiToggle>
</styled.div>
`;
